core/c4diagram.go

```diff
@@ -0,0 +1,355 @@
+func encode64(e []byte) string {
+	var r bytes.Buffer
+	for i := 0; i < len(e); i += 3 {
+		switch len(e) {
+		case i + 2:
+			r.Write(append3bytes(e[i], e[i+1], 0))
+		case i + 1:
+			r.Write(append3bytes(e[i], 0, 0))
+		default:
+			r.Write(append3bytes(e[i], e[i+1], e[i+2]))
+		}
+	}
+	return r.String()
+}
+
+func append3bytes(e, n, t byte) []byte {
+	c1 := e >> 2
+	c2 := (3&e)<<4 | n>>4
+	c3 := (15&n)<<2 | t>>6
+	c4 := 63 & t
+
+	var buf bytes.Buffer
+
+	buf.WriteByte(encode6bit(c1 & 63))
+	buf.WriteByte(encode6bit(c2 & 63))
+	buf.WriteByte(encode6bit(c3 & 63))
+	buf.WriteByte(encode6bit(c4 & 63))
+
+	return buf.Bytes()
+}
+
+func encode6bit(e byte) byte {
+	if e < 10 {
+		return 48 + e
+	}
+
+	e -= 10
+	if e < 26 {
+		return 65 + e
+	}
+
+	e -= 26
+	if e < 26 {
+		return 97 + e
+	}
+
+	e -= 26
+	switch e {
+	case 0:
+		return '-'
+	case 1:
+		return '_'
+	default:
+		return '?'
+	}
+}
+
+func trimQuotes(s string) string {
+	if s[:1] == `"` {
+		s = s[1:]
+	}
+	if s[len(s)-1:] == `"` {
+		s = s[:len(s)-1]
+	}
+	return s
+}
+
+func diagramFooter2UML(graph DiagramGraph) string {
+	if graph.Footer == "" {
+		return `footer "generated by diagramastext.dev - %date('yyyy-MM-dd')"`
+	}
+	return `footer "` + stringCleaner(graph.Footer) + `"`
+}
+
+func diagramNode2UML(n *Node) (string, error) {
+	if n.ID == "" {
+		return "", errors.New("node must be identified: 'id' attribute")
+	}
+
+	o := "Container"
+
+	switch n.IsQueue && n.IsDatabase {
+	case true:
+	case false:
+		if n.IsQueue {
+			o += "Queue"
+		}
+
+		if n.IsDatabase {
+			o += "Db"
+		}
+	}
+
+	if n.External {
+		o += "_Ext"
+	}
+
+	o += "(" + n.ID
+
+	label := n.Label
+	if label == "" {
+		label = n.ID
+	}
+	o += `, "` + stringCleaner(label) + `"`
+
+	if n.Technology != "" {
+		o += `, "` + stringCleaner(n.Technology) + `"`
+	}
+
+	o += ")"
+
+	return o, nil
+}
+
+func diagramLink2UML(l *Link) (string, error) {
+	if l.From == "" || l.To == "" {
+		return "", errors.New("link must specify the end nodes: 'from' and 'to' attributes")
+	}
+
+	o := "Rel"
+
+	if d := linkDirection(l.Direction); d != "" {
+		o += "_" + d
+	}
+
+	o += "(" + l.From + ", " + l.To
+
+	if l.Label != "" {
+		o += `, "` + stringCleaner(l.Label) + `"`
+	}
+
+	if l.Technology != "" {
+		o += `, "` + stringCleaner(l.Technology) + `"`
+	}
+
+	o += ")"
+
+	return o, nil
+}
+
+func linkDirection(s string) string {
+	switch s := strings.ToUpper(s); s {
+	case "LR":
+		return "R"
+	case "RL":
+		return "L"
+	case "TD":
+		return "D"
+	case "DT":
+		return "U"
+	default:
+		return ""
+	}
+}
+
+func stringCleaner(s string) string {
+	s = strings.TrimSpace(s)
+	s = trimQuotes(s)
+	s = strings.ReplaceAll(s, "\n", "\\n")
+	return s
+}
+
+// diagramGraph2plantUMLCode function to "transpile" the diagram definition graph to plantUML code as string.
+func diagramGraph2plantUMLCode(graph DiagramGraph) (string, error) {
+	if len(graph.Nodes) == 0 {
+		return "", errors.New("graph must contain at least one node")
+	}
+
+	o := `@startuml
```

Comment:
**suggestion** use `strings.Builder` to create this string efficiently.

Reply:
For refactoring